Merge develop into experimental client

sechkova opened the pull request:
**Description of the changes being introduced by the pull request**:

Merge develop to get latest changes.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
New API: Add root use case in couple of tests

Add a use case for the root class to be tested in test_generic_read
and test_read_write_read_compare tests in test_apy.py

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
NEW API: Fix documentation indentation

From the reST/sphinx docs:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

I added new lines and an identation where it was missed.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Make length and hashes optional in Timestamp

As per the specification (v1.0.1) length and hashes fields
in timestamp and snapshot metadata are optional.
We have implement this in the older API
(see https://github.com/theupdateframework/tuf/pull/1031) and we should
implement it in the new API.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add Key class and integrate it into Root

In the top level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add a Role class and integrate it into Root

In the top level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add DelegatedRole and Delegations

In the top level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

DelegatedRole shares a couple of fields with the Role class and that's
why it inherits it.
I decided to use a separate Delegations class because I thought it will
make it easier to read, verify and add additional helper functions.
Also, I tried to make sure that I test each level of the delegations
representation for support of storing unrecognized fields.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add from_dict methods in Key and Role classes

The from_dict() method simplifies the object creation.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Make keyids in Role a set

From the specification:
"Clients MUST ensure that for any KEYID represented in this key list
and in other files, only one unique key has that KEYID."

The “only one unique key has that KEYID” is a requirement which can’t
be achieved if two keyids are the same.
So, in order to mandate that requirement it makes sense to use a set
which will guarantee us the keyid’s uniqueness.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Check the format of keyval in Root

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
New API: Add Key/Role specific tests

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
tox: Don't use pre-release tools

Keep using newest versions of build tools, but don't use pre-releases:
* less downloading (with chances of failures) happens in the CI because
  versions change less often
* tools are less likely to break the build (or at least it happens less
  often)

This change should not affect the software under test as we install
pinned versions of those.

Note that this also doesn't affect black: tox still ends up with a
pre-release version of black because that project has never made an
actual release.

Fixes #1350

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
keyid: verify adding an existing key is ignored

Verify that adding an already existing key to keyid for a particular
role in Root won't create duplicate key.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
BREAKING CHANGE: Make delegations optional

According to the spec, delegations in targets are marked as optional:
https://theupdateframework.github.io/specification/latest/#file-formats-targets
and a pr, clarifying that even more, is approved:
https://github.com/theupdateframework/specification/pull/157.

This is a possible breaking change.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Merge pull request #1360 from MVrachev/key-role-classes

New API: Add Key and Role classes

sechkova committed:
CI: Allow failure when publishing on coveralls

A failure during publishing of the coverage results
on coveralls should not fail the whole build job.
Allow the step to fail.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

Commit:
Add Delegation/DelegatednRole specific tests

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
API: Make _type read-only

Also remove _type from Signed constructor arguments: the value is in a
class atttribute. This way _type never needs to be validated (except
in the dispatcher in Metadata). There is a double-check in
_common_fields_from_dict() just to be sure.

This makes the API easier to use correctly as the public property is
immutable.

This is an API break as all Signed constructors change -- this could be
avoided but seems like the correct choice.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Merge pull request #1369 from sechkova/coveralls-publish

CI: Allow failure when publishing on coveralls

Commit:
Merge pull request #1364 from jku/no-pre-release-tools

Commit:
Merge pull request #1372 from jku/read-only-type

API: Make _type read-only

Commit:
Metadata API: Provide type as well as _type

Using Metadata APIs '_type' attribute (from outside metadata.py) currently
results in linter errors.

Add a duplicate 'type': this way the API users can avoid linter errors
but '_type' is still available in case the strict file format
compatibility is needed.

Fixes #1375

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: remove_key() improvements

We should not do multiple lookups through data structures if one is
enough (here we have extra lookups on both roles and keyids).

Also in this case raising on missing key seems like the preferable
alternative so even a try-except is not needed.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Fix type hints

* Define missing argument type hints
* Stop using Mapping where we actually mean Dict:
  Mapping means "we only need a read-only dict" and most of the
  time this is not really the case.
* Use List, not list (latter only works from Python 3.9)
* Update Metadata.signatures documentation

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
build(deps): bump securesystemslib

Bumps [securesystemslib[crypto,pynacl]](https://github.com/secure-systems-lab/securesystemslib) from 0.20.0 to 0.20.1
- [Release notes](https://github.com/secure-systems-lab/securesystemslib/releases)
- [Changelog](https://github.com/secure-systems-lab/securesystemslib/blob/master/CHANGELOG.md)
- [Commits](https://github.com/secure-systems-lab/securesystemslib/compare/v0.20.0...v0.20.1)

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Merge pull request #1381 from jku/bump-sslib

build(deps): bump securesystemslib

Commit:
Merge pull request #1367 from MVrachev/improvements

Make hashes, length and delegations optional + improvements

Commit:
Remove forgotten debug line from a test

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Don't use Mapping for mutable dicts

Stop using Mapping where we actually mean Dict:
Mapping means "we only need a read-only dict" and most of the time
this is not really the case.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

joshuagl committed:
Merge pull request #1376 from jku/duplicate-type-property

Metadata API: Provide type as well as _type

Commit:
Metadata API/pylintrc: Use old style logging

This is suggested by the Google style guide: the old style logging
(%-format) allows the log strings to be lazily formatted so there's less
need to think about performance when forming debug messages.

No actual code changes are needed because the metadata API does not yet
log anything.

Fixes #1334

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Merge pull request #1383 from jku/old-style-logging

Metadata API/pylintrc: Use old style logging

Commit:
Merge pull request #1379 from jku/metadata-minor-fixes

Metadata minor fixes

Commit:
Merge pull request #1370 from MVrachev/delegation-classes

New API: Add DelegationRole and Delegations classes

sechkova committed:
Merge branch 'develop' into experimental-client

sechkova committed:
Fix issues after merging develop

Fix failing tests after merging develop into
experimental-client.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

jku commented on tuf/client_rework/updater_rework.py:
```diff
@@ -558,7 +558,7 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
             number_of_delegations -= 1
             delegations = role_metadata.delegations
-            child_roles = delegations.get("roles", [])
+            child_roles = delegations.roles
```
delegations can be none

sechkova replied:
Added one additional commit addressing this.

jku commented on tuf/client_rework/metadata_wrapper.py:
```diff
@@ -95,8 +95,8 @@ def keys(self, role):
-        for keyid in self._meta.signed.roles[role]["keyids"]:
-            key_metadata = self._meta.signed.keys[keyid]
+        for keyid in self._meta.signed.roles[role].keyids:
+            key_metadata = self._meta.signed.keys[keyid].to_dict()
```
This is fine here but ... for future I guess we want to modify the Metadata.verify() function so that it takes a Key as argument so that this _TUF Key -> dictionary ->  SSLIB Key_ dance isn't needed ? If you agree, can you add that as comment in the already open verification API improvement issue or file a new issue about it?

sechkova replied:
Mentioned it [here](https://github.com/theupdateframework/tuf/issues/1377#issuecomment-840444942).

jku commented on tuf/client_rework/metadata_wrapper.py:
```diff
@@ -158,20 +158,20 @@ def keys(self, role):
-                for keyid in delegation["keyids"]:
-                    key_metadata = self._meta.signed.delegations["keys"][keyid]
-                    key, dummy = format_metadata_to_key(key_metadata)
+        for delegation in self._meta.signed.delegations.roles:
```
delegations can be none

jku commented on tuf/client_rework/metadata_wrapper.py:
```diff
@@ -158,20 +158,20 @@ def keys(self, role):
-        for delegation in self._meta.signed.delegations["roles"]:
-            if delegation["name"] == role:
-                return delegation["threshold"]
+        for delegation in self._meta.signed.delegations.roles:
```
delegations can be none

jku reviewed (commented):
The comments are mostly for the wrapper (so not super important as it will be replaced by metadatabundle) but I do appreciate it if you solve these so I'll have less work in a few days when I update the bundle PR :grimacing: 

Most comments are about the annoying feature in Metadata API where some containers (like delegations) can be None (and not just empty) because it signifies that the actual item is missing in the file format.  Often you can probably do `delegations = signed.delegations or []`

sechkova committed:
Prevent access to missing metadata fields

Delegations are optional. Add checks to avoid accessing
'Delegations' object's members if its value is None.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>
